[sw] Add Guidance on DIF Return Codes

lenary opened the pull request:
This PR adds the guidance as agreed in the software meeting on 2020-04-14. The guidance is also based on the discussion in #1911. I have also updated the style guide to refer to enum constants by that name (which is used by the standard), rather than "enum variants", which is not.

I have also added `bitfield.h` as an allowed library for DIFs, given it is already used and #2203 is likely to use it even more. 

---

Closes #1911

lenary commented:
I changed the order of the return codes to more accurately reflect the order in existing DIFs.

jon-flatley commented on sw/device/lib/dif/README.md:
```diff
@@ -50,6 +51,30 @@ there are some relaxations of these rules for them described at the end.
+    the three general constants.
+  * DIFs must document the meaning of all error codes, including to what extent
+    the error is recoverable, or to what extent the operation is retryable.
+    `kDif<ip>Error` is not recoverable. `kDif<ip>BadArg` is designed to be
```
Is `kDif<ip>BadArg` really recoverable? To me this implies that the caller unknowingly supplied an invalid argument, and therefore represents a bug in the implementation of the driver supplying the arguments to the respective DIF. It seems to me that trying to implement recovery for a bug that the driver itself caused only increases complexity, and I can't imagine a test case for this code path that wouldn't catch the bug causing the `BadArg` in the first place.

silvestrst replied:
I agree, not sure if `kDif<ip>BadArg` is recoverable. If you pass it `null`, you presumably didn't do it on purpose, and there is not much you can do about it.

However, if you add meaningful error codes to functions like `dif_uart_configure`, you possibly can recover from things like `kDifUartBadNco` **(just an example)**.

alphan replied:
I am guessing unrecoverable errors will result in some drastic action. If so, it is reasonable to classify `kDif<ip>BadArg` as recoverable, especially in a shared context. Consider the following:
- There are multiple apps, using the same device.
- They share their buffers with the device DIF library (probably via the device capsule in tock, not clear if the capsule will perform the same checks as DIF libraries do, I am assuming it doesn't but even if it does the same discussion would apply to it too)
- BadApp has a bug and passes a `NULL` pointer.

Even if BadApp cannot meaningfully retry the call, we may want to let the rest of the system keep going. I'm not sure what happens to BadApp though (log, restart/kill, and obviously fix?).

That being said, I also understand the other point of view. BadApp is there for a reason and it cannot perform its function if it passes a `NULL` pointer and restarting it may just result in a restart loop.

GregAC replied:
I think it's reasonable to document `kDif<ip>BadArg` as recoverable. It's up to the SW using the DIFs whether or not it uses that property but from a documentation point of view it's good to know what is and isn't possible.

lenary replied:
Ok, so it sounds like it's good to document the property "this return code means the device has been left in an improper (or unknown) state" vs "this return code means the operation was not performed, but/and the device is still in a usable state". 

It seems "Recoverable" isn't the right word for this, but that property is at least useful to document. I will try to work out a way of phrasing this.

mcy replied:
I've felt that all errors can be given one of a few "remediation strategies":
- Reboot the device. You're in an impossible situation.
- Retry with different arguments (in a lot of cases, this degenerates to "reboot the device").
- Retry with the same arguments. (E.g., you got quota'd or evicted.)

I think that `Error` and `BadArg` correspond to the first two, respectively, while you could imagine IPs which have errors with the third kind. I think a better phrasing might be along the lines of: "Every error must specifiy a remediation strategy, such as [...]." We should perhaps be careful as to what kinds of remediation strategies we allow, to avoid overly complicated caller-side control flow based on errors.

jon-flatley commented on sw/device/lib/dif/README.md:
```diff
@@ -50,6 +51,30 @@ there are some relaxations of these rules for them described at the end.
+  * DIF libraries must have a general return code enum called
+    `dif_<ip>_result_t`. It must define the following constants (in this order):
+    * `kDif<ip>Ok`, with value 0, to denote the call succeeded.
+    * `kDif<ip>Error`, to denote a non-specific error happened during the call.
```
Is `Error` the best term here? After all `BadArg` is also an error. Maybe something like `UnknownError` or just `GeneralError`? Or is the goal here to decrease verbosity?

silvestrst replied:
I think `kDif<ip>Error` is fine, we discussed it #1911 . In my opinion all of these variants are fine, so we might just keep the present one.

alphan replied:
While we are at it, maybe we should remove individual `dif_<ip>_result_t`s and define a common `dif_result_t`. What do you think?

lenary replied:
The idea of a single `dif_result_t` was floated at the call, and I am very against it, because it would have to live in a centralised header, and I do not want more dependencies for the DIFs. 

We're trying to ensure the DIFs remain reusable on other projects and so can be individually vendored like the peripheral RTL can be vendored. We cannot really avoid the use of the `mmio` library (and the like), but I want to ensure we keep other dependencies to a minimum. I don't see the payoff for sharing an enum like this yet.

lenary replied:
I think it is prudent to give explicit values to the `dif_<ip>_result_t` constants, and any in more specialised enums (so that the `Ok`, `Error`, and `BadArg` values match).

mcy replied:
I think the "main" payoff is being able to write something like `RETURN_IF_ERR` or like Rust's `?` operator. However, it is still unclear if this has that much utility, and we can simply mandate that `Ok` always has the representation `0`.

jon-flatley commented on sw/device/lib/dif/README.md:
```diff
@@ -50,6 +51,30 @@ there are some relaxations of these rules for them described at the end.
+    the three constants in `dif_<ip>_result_t`.
+    The additional, specific return code constants must all be defined after
+    the three general constants.
+  * DIFs must document the meaning of all error codes, including to what extent
```
The language here makes it unclear to me if this includes `kDif<ip>Ok`, `kDif<ip>Error`, and/or `kDif<ip>BadArg` or just the error codes belonging to `dif_<ip>_<operation>_result_t`.

silvestrst commented on sw/device/lib/dif/README.md:
```diff
@@ -51,6 +51,30 @@ there are some relaxations of these rules for them described at the end.
+    multiple functions that fail in the same way.
+    These more specific return codes must be named
+    `dif_<ip>_<operation>_result_t`, their constants should be prefixed with
+    `kDif<ip><operation>`, and the first three constants should be the same as
+    the three constants in `dif_<ip>_result_t`.
```
Maybe something like: "`kDif<ip><operation>` must define all of `dif_<ip>_result_t` variants ..." ?

silvestrst replied:
Just a nit, im happy the way it is! :)

alphan commented:
Posting on the main thread for visibility:

I think we should define values of all variants of `dif_ip_result_t` explicitly not just `kDifIpOk` to make sure that they are compatible across all functions (they are shared return codes after all):
```
typedef enum dif_ip_result {
  kDifIpOk = 0,
  kDifIpError = 1,
  kDifIpBadArg = 2,
} dif_ip_result_t;
```
and
```
typedef enum dif_ip_op_result {
  kDifIpOpOk = kDifIpOk,
  kDifIpOpError = kDifIpError,
  kDifIpOpBadArg = kDifIpBadArg,
  kDifIpOpSomeError,
  kDifIpOpAnotherError,
  ...
} dif_ip_op_result_t;
```

This may be a minor point but the guidance doesn't specify their values in its current form.

lenary commented on sw/device/lib/dif/README.md:
```diff
@@ -50,6 +51,30 @@ there are some relaxations of these rules for them described at the end.
+      This is for the `default:` case of enum switches (as noted below), and for
+      assertion errors (usually where the function has already caused side
+      effects so `kDif<ip>BadArg` cannot be used).
+    * `kDif<ip>BadArg`, to denote that the caller supplied incorrect arguments.
```
Given what we've done in the UART DIF for configuration, my feeling is if there's a more specific reason an argument is bad (in the UART example, you may have chosen values that set the NCO to an unrepresentable value), then you can return a more specific result constant.

However, I think it is good to always have `BadArg` in the enums, so that things like passing a NULL `dif_<ip>_t*` (which usually includes the base address of the device) can still use that return code (given it will in the generic functions).

jon-flatley commented:
> This may be a minor point but the guidance doesn't specify their values in its current form.

I think it does, given the two requirements 

> It must define the following constants (in this order)

 and 

> the first three constants should be the same as the three constants in `dif_<ip>_result_t`

I suppose the language doesn't specifically speak to their values, so maybe you could do something silly like 

```
typedef enum dif_ip_op_result {
  kDifIpOpOk = 1,
  kDifIpOpError,
  kDifIpOpBadArg,
  kDifIpOpSomeError,
  kDifIpOpAnotherError,
  ...
} dif_ip_op_result_t;
```

I do like the explicit valuing of these first three constants. It makes it clear that comparing these values across all `dif_<ip>_<op>_result_t` types is valid, assuming that's the intent.

alphan commented:
> It makes it clear that comparing these values across all `dif_<ip>_<op>_result_t` types is valid.

You are right, this was the point I was trying to make.

mcy commented on sw/device/lib/dif/README.md:
```diff
@@ -50,6 +51,30 @@ there are some relaxations of these rules for them described at the end.
+    * `kDif<ip>BadArg`, to denote that the caller supplied incorrect arguments.
+      This value must only be returned before the function does anything that
+      causes side effects.
+  * Other return code enums can be defined for functions that fail in more
+    specific ways, and these other return code enums can be shared between
```
To be clear, `<operation>` is any arbitrary string, and need not correspond to any single function? I think this is the most reasonable reading, but I want to be sure.

mcy commented on sw/device/lib/dif/README.md:
```diff
@@ -61,16 +86,18 @@ there are some relaxations of these rules for them described at the end.
-    not an enum variant.
-  * Enum switches do not need a `case` for enum variants that are unreachable
+    not a constant from that enum. In the absence of more specific information,
+    this should return `kDif<ip>Error` or the equivalent return code value from
+    a more specific return code enum.
```
Is `BadArg` acceptable if this switch happens before any side-effects occur?

lenary commented:
Ok, I think I have addressed all the comments and questions in this guidance. If you had comments, please re-check what I have written.

Other changes I have added to this PR:
- Mandated DIF headers be "polyglot headers". This is needed because we use them from both C and C++.
- Downgraded the requirements on not checking enum argument ranges to `may not` instead of `should not`. Hopefully this avoids some of the confusion around this rule and the rule for `default:` in enum switches being slightly contradictory, feedback I got from @GregAC this week.

lenary commented:
As I said in the OT Software meeting, it would be useful if this could be re-reviewed so that it can be merged and I can start a new batch of changes relating to issues that have come out of the first few Design Reviews.

I'm now going to set a deadline, as it's already been about a week since my last comment. I would like any further comments by the end of Friday 12th June (AoE).

mcy commented on sw/device/lib/dif/README.md:
```diff
@@ -60,17 +108,17 @@ there are some relaxations of these rules for them described at the end.
-    not an enum variant.
-  * Enum switches do not need a `case` for enum variants that are unreachable
+    not a constant from that enum. In the absence of more specific information,
+    this must return `kDif<ip>Error` or the equivalent return code value from
+    a more specific return code enum.
```
Nit: include a caveat that `kDif<ip>BadArg` may be returned if the switch is part of a precondition check.

lenary replied:
Yeah, re-reading this it shouldn't be a `must` sentence.

mcy commented on sw/device/lib/dif/README.md:
```diff
@@ -48,9 +58,47 @@ there are some relaxations of these rules for them described at the end.
+    return codes defined above, including more specific reasons arguments are
+    invalid.
+
+  * DIFs must document the meaning of all return codes, including the required
```
Nit: I assume this means that the return code enum must contain doxygen comments for all of its variants; maybe we should clarify this?

(I could imagine someone thinking that it would need to be documented on the DIF's `@return` comment instead...)

alphan commented on sw/device/lib/dif/README.md:
```diff
@@ -31,16 +31,26 @@ contradict with the guidelines below.
+### Definitions
+
+<a name="side-effects"></a>
+**Side-effects** include (but are not limited to) writing to memory, including
+memory-mapped hardware, and modifying processor CSRs. This means returning early
+from a guard statement must not leave the hardware in an invalid or
+unrecoverable state.
```
This is really minor but I am wondering if the second sentence should be more specific. What do you think about the following?

"If the state of the hardware, i.e. the processor and all peripherals, is _S_ at the beginning of the function and _S'_ at the point of early return, then _S_ and _S'_ can differ only in those components that change indirectly due to the code that executes until the early return, e.g. timer counters, program counter, etc."

lenary replied:
Two things:
- I think that the second sentence here is actually not part of the definition of side effects, it's about guard statements, and I got it in the wrong place, I'll push it further down the file again (to the rules of guard statements).
- I don't want to be as formal as your proposed sentence - I don't think it helps to think about this abstractly. I prefer to talk about the kinds of actions that are considered to change peripherals in a way that we don't want, rather than whitelisting a set of actions that are allowed.

alphan replied:
> * I think that the second sentence here is actually not part of the definition of side effects, it's about guard statements, and I got it in the wrong place, I'll push it further down the file again (to the rules of guard statements).

This clears things up, thanks!

lenary replied:
The most recent update was to address the issue with the side-effects definition.

lenary commented on sw/device/lib/dif/README.md:
```diff
@@ -31,16 +31,26 @@ contradict with the guidelines below.
 
+### Definitions
+
+<a name="side-effects"></a>
```
I don't want the anchor to go with the h3, and boldening stuff does not create an anchor.

gkelly replied:
Makes sense, thanks!

lenary commented:
I'll merge this later today.
